ext/js/language/translator.js

```diff
@@ -514,26 +539,45 @@ export class Translator {
      * @param {import('language').TextProcessorWithId<unknown>[]} textProcessors
      * @param {(import('translation').FindTermsTextReplacement[] | null)[]} textReplacements
      * @param {import('translation-internal').TextCache} textCache
-     * @returns {Set<string>}
+     * @returns {import('translation-internal').VariantAndTextProcessorRuleChainCandidatesMap}
      */
     _getTextVariants(text, textProcessors, textReplacements, textCache) {
-        let variants = new Set([text]);
+        // let variants = new Set([text]);
+
+        /** @type {import('translation-internal').VariantAndTextProcessorRuleChainCandidatesMap} */
+        let variantsMap = new Map([
+            [text, [[]]],
+        ]);
+
         for (const textReplacement of textReplacements) {
             if (textReplacement === null) { continue; }
-            variants.add(this._applyTextReplacements(text, textReplacement));
+            variantsMap.set(this._applyTextReplacements(text, textReplacement), [[]]);
         }
-        for (const {id, textProcessor: {process, options}} of textProcessors) {
-            /** @type {Set<string>} */
-            const newVariants = new Set();
-            for (const variant of variants) {
+        for (const {id, textProcessor: {name, process, options}} of textProcessors) {
+            /** @type {import('translation-internal').VariantAndTextProcessorRuleChainCandidatesMap} */
+            const newVariantsMap = new Map();
+            for (const [variant, currentPreprocessorRuleChainCandidates] of variantsMap) {
                 for (const option of options) {
                     const processed = this._getProcessedText(textCache, variant, id, option, process);
-                    newVariants.add(processed);
+                    const existingCandidates = newVariantsMap.get(processed);
+
+                    // Ignore if applying the textProcessor doesn't change the source
+                    if (processed === variant) {
+                        if (typeof existingCandidates === 'undefined') {
+                            newVariantsMap.set(processed, currentPreprocessorRuleChainCandidates);
+                        } else {
+                            newVariantsMap.set(processed, existingCandidates);
+                        }
+                    } else if (typeof existingCandidates === 'undefined') {
+                        newVariantsMap.set(processed, currentPreprocessorRuleChainCandidates.map((candidate) => [...candidate, name]));
+                    } else {
+                        newVariantsMap.set(processed, [...existingCandidates, ...currentPreprocessorRuleChainCandidates.map((candidate) => [...candidate, name])]);
+                    }
```

Comment:
```suggestion
                    } else if (typeof existingCandidates === 'undefined') {
                        newVariantsMap.set(processed, currentPreprocessorRuleChainCandidates.map((candidate) => [...candidate, id]));
                    } else {
                        newVariantsMap.set(processed, [...existingCandidates, ...currentPreprocessorRuleChainCandidates.map((candidate) => [...candidate, id])]);
                    }
```

Using ID here might be better, since we uniqueness of names isn't enforced